napari/layers/image/experimental/octree.py

```diff
@@ -67,240 +74,285 @@ def _create_tiles(array: np.ndarray, tile_size: int) -> np.ndarray:
+            for col in col_range:
 
-    Parameters
-    ----------
-    levels : Levels
-        Print information about these levels.
-    """
-    print(f"{len(levels)} levels:")
-    for level in levels:
-        _print_level_tiles(level)
+                data = self.tiles[row][col]
+                pos = [x, y]
 
+                if 0 not in data.shape:
+                    chunks.append(ChunkData(data, pos, scale_vec))
 
-def _print_tiles(node, level=0):
-    assert node is not None
-    assert node.tile is not None
-    node.print_info(level)
-    for child in node.children:
-        if child is not None:
-            _print_tiles(child, level + 1)
+                x += data.shape[1] * scale
+            y += data.shape[0] * scale
 
+        return chunks
 
-class OctreeNode:
-    """Octree Node.
+    def tile_range(self, span, num_tiles):
+        """Return tiles indices needed to draw the span."""
 
-    Child indexes
-    -------------
-    OCTREE_TODO: This order was picked arbitrarily. If there is another
-    ordering which makes more sense, we should switch to it.
+        def _clamp(val, min_val, max_val):
+            return max(min(val, max_val), min_val)
 
-    -Z [0..3]
-    +Z [4..7]
+        tile_size = self.info.tile_size
 
-        -X X+
-    -Y 0 1
-    +Y 3 2
+        tiles = [span[0] / tile_size, span[1] / tile_size]
+        new_min = _clamp(tiles[0], 0, num_tiles - 1)
+        new_max = _clamp(tiles[1], 0, num_tiles - 1)
+        clamped = [new_min, new_max + 1]
 
-        -X X+
-    -Y 4 5
-    +Y 7 6
+        span_int = [int(x) for x in clamped]
+        return range(*span_int)
 
-    Parameters
-    ----------
-    row : int
-        The row of this octree node in its level.
-    col : int
-        The col of this octree node in its level.
-    data : np.ndarray
-        The image data for this octree node.
-    """
+    def row_range(self, span):
+        """Return row indices which span image coordinates [y0..y1]."""
+        return self.tile_range(span, self.num_rows)
 
-    def __init__(self, row: int, col: int, data: np.ndarray):
-        assert data is not None
-        assert row >= 0
-        assert col >= 0
-        self.row = row
-        self.col = col
-        self.data = data
-        self.children = None
+    def column_range(self, span):
+        """Return column indices which span image coordinates [x0..x1]."""
+        return self.tile_range(span, self.num_cols)
 
-    def print_info(self, level):
-        """Print information about this octree node.
 
-        level : int
-            The level of this node in the tree.
-        """
-        indent = "    " * level
-        print(
-            f"{indent}level={level} row={self.row:>3}, col={self.col:>3} "
-            f"shape={self.tile.shape}"
-        )
+def _one_tile(tiles: TileArray) -> bool:
+    return len(tiles) == 1 and len(tiles[0]) == 1
 
 
 class Octree:
-    """An octree.
+    """A region octree that holds hold 2D or 3D images.
```

Comment:
Can be a discussion for another time, but I'm curious how you see this working eventually with 4D or 5D data, say mulitcolor volumetric timeseries? Do the first two, non-spatial dimensions (color, time) just get sliced/ index in a non multiresolution way with a chunk size of (1, 1, 64, 64, 64)? Don't need to worry about now, but we know that we'll want to support data like that eventually

Reply:
Even for 2D timeseries there's a choice. It's either an array of quadtrees, or it's a quadtree where every node is a timeseries array of just that tile. This is the classic "array of structures or a structure of arrays" type of thing.

I think in many cases "array of structures" is your first guess, but "structure of arrays" is very often more efficient, and I think it's more efficient here. And that's what we did for Raveller. 

The only tweak is the nodes don't contain arrays they contained dicts, so like sparse arrays. So a node might only have data for 1 slice in it. Or it might have data for all the slices if you loaded them all. However, that's basically a type of caching so it would have to work with the cache, you'd have to consider where you want to cache stuff.

For example if you have a giant stack of images and you zoom way in and then go from slice-to-slice following something, say you follow it through 100 slices. Maybe you are viewing just 12 tiles, so now those 12 nodes in the quadtree each contain a dict of 100 slices.

For 3D time series I think it's the same. So now it's a true octree, but each node is a dict/timeseries for what's loaded at that level of detail in that location. So it's a dict of sub-volumes.

Color or other channels I'd also put down into the nodes also. We did have 2 channels for Raveller the grayscale image and an RGB-style superpixel index image. So we did that and it worked.

It requires more thought though to see of napari really should do it the same way or not. I have not written anything in either direction yet. Just have not gotten there yet, but might in async-16 or 17.

Reply:
And maybe there is where we consider using dask deep in the internals, such that everything is dask regardless of what it started as. That may or may not be a good idea. I don't think it's obvious either way.